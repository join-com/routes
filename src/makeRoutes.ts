import { RoutesConfig, Routes, Options } from './types'

import buildRoute from './buildRoute'

const routeNames = (config: RoutesConfig) => [
  ...new Set(config.routes.map(({ name }) => name))
]

const makeRoutes = <T extends string>(config: RoutesConfig): Routes<T> => {
  const routes = (routeName: T) => (options?: Options) =>
    buildRoute(config, routeName, options)

  routeNames(config).forEach(routeName => {
    routes[routeName] = (options?: Options) =>
      buildRoute(config, routeName, options)
  })

  return routes as Routes<T>
}

export default makeRoutes
