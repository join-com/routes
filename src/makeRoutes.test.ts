import makeRoutes from './makeRoutes'

describe('makeRoutes', () => {
  const routeParams = {
    id: '1-job-title',
    companyId: '2-company-name'
  }

  type RouteName = 'root' | 'showJob' | 'showJobSubdomain' | 'featuresManage'

  const defaultConfig = {
    locale: 'en-us',
    baseUrl: 'http://example.com:3000',
    routes: [
      {
        name: 'root',
        locale: 'en-us',
        pattern: '/'
      },
      {
        name: 'root',
        locale: 'de-de',
        pattern: '/de-de'
      },
      {
        name: 'showJob',
        pattern: '/companies/:companyId/jobs/:id'
      },
      {
        name: 'showJobSubdomain',
        pattern: '/jobs/:id',
        options: { subdomain: true }
      },
      {
        name: 'featuresManage',
        pattern: '/features/ATS-applicant-management'
      },
      {
        name: 'featuresManage',
        pattern: '/de-de/features/bewerber-verwalten-bms',
        locale: 'de-de'
      },
      {
        name: 'featuresManage',
        pattern: '/de-ch/features/bewerber-verwalten-bms',
        locale: 'de-ch'
      }
    ]
  }

  const appRoute = makeRoutes<RouteName>(defaultConfig)

  const testRouteUrl = (
    routes: any,
    routeName: string,
    params: any,
    expectedUrl: string
  ) => {
    expect(routes(routeName)(params)).toEqual(expectedUrl) // old signature
    expect(routes[routeName](params)).toEqual(expectedUrl) // new signature
  }

  const testException = (routes: any, routeName: string, params: any) => {
    try {
      routes(routeName)(params)
    } catch (error) {
      expect(error).toMatchSnapshot()
    }

    try {
      routes[routeName](params)
    } catch (e) {
      expect(e).toMatchSnapshot()
    }
  }

  it('has route name type support', () => {
    appRoute('featuresManage')({})
    appRoute.featuresManage({})

    expect.assertions(0)
  })

  it('generate route for specified locale', () => {
    testRouteUrl(
      appRoute,
      'featuresManage',
      { locale: 'de-ch' },
      'http://example.com:3000/de-ch/features/bewerber-verwalten-bms'
    )

    expect.assertions(2)
  })

  it('applies route params', () => {
    testRouteUrl(
      appRoute,
      'showJob',
      routeParams,
      'http://example.com:3000/companies/2-company-name/jobs/1-job-title'
    )

    expect.assertions(2)
  })

  it('adds remain params to query', () => {
    testRouteUrl(
      appRoute,
      'showJob',
      {
        ...routeParams,
        key1: 'value1',
        key2: 'value2'
      },
      'http://example.com:3000/companies/2-company-name/jobs/1-job-title?key1=value1&key2=value2'
    )

    expect.assertions(2)
  })

  it('raises error when missing params', () => {
    const paramsMissingCountry = {
      id: '1-job-title'
    }

    testException(appRoute, 'showJob', paramsMissingCountry)

    expect.assertions(2)
  })

  it('raises error when route can not be found', () => {
    const options = {
      locale: 'ua'
    }

    testException(appRoute, 'featuresManage', options)

    expect.assertions(2)
  })

  it('generate route for fallback locale if specified locale was not found', () => {
    testRouteUrl(
      makeRoutes({ ...defaultConfig, fallbackLocale: 'en-us' }),
      'featuresManage',
      { locale: 'it-fr' },
      'http://example.com:3000/features/ATS-applicant-management'
    )

    expect.assertions(2)
  })

  describe('when route requires subdomain', () => {
    it('raises error when subdomain params is missed', () => {
      const params = {
        id: '1-job-title'
      }

      testException(appRoute, 'showJobSubdomain', params)

      expect.assertions(2)
    })

    it('applies route params', () => {
      const param = {
        id: '1-job-title',
        subdomain: 'test'
      }

      testRouteUrl(
        appRoute,
        'showJobSubdomain',
        param,
        'http://test.example.com:3000/jobs/1-job-title'
      )
    })
  })
})
