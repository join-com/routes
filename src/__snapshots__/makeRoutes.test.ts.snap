// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeRoutes raises error when missing params 1`] = `[TypeError: Expected "companyId" to be a string]`;

exports[`makeRoutes raises error when missing params 2`] = `[TypeError: Expected "companyId" to be a string]`;

exports[`makeRoutes raises error when route can not be found 1`] = `[Error: Route featuresManage with locale ua does not exist.]`;

exports[`makeRoutes raises error when route can not be found 2`] = `[Error: Route featuresManage with locale ua does not exist.]`;

exports[`makeRoutes when route requires subdomain raises error when subdomain params is missed 1`] = `[Error: Expected "subdomain" to be defined]`;

exports[`makeRoutes when route requires subdomain raises error when subdomain params is missed 2`] = `[Error: Expected "subdomain" to be defined]`;
